scaler.rb

```diff
@@ -36,28 +34,82 @@
   },
 }
 
-updates = {
+config_updates = {}
+
+formation_updates = {
   "process" => options[:process],
   "quantity" => options[:quantity]
 }
 
-config_updates = {}
+preboot_enabled   = heroku.app_feature.info(options[:app], 'preboot')["enabled"]
+current_formation = heroku.formation.info(options[:app], options[:process])
+current_config    = heroku.config_var.info(options[:app])
 
-if options[:size]
-  options[:size].upcase!
-  updates["size"] = options[:size]
-  if default_config[options[:size]]['RUBY_GC_HEAP_GROWTH_MAX_SLOTS']
-    config_updates['RUBY_GC_HEAP_GROWTH_MAX_SLOTS'] = default_config[options[:size]]['RUBY_GC_HEAP_GROWTH_MAX_SLOTS']
-  else
-    config_updates['RUBY_GC_HEAP_GROWTH_MAX_SLOTS'] = nil
+# Default to updating config first, then formation
+commands = [:config, :formation]
+
+if options[:web_concurrency]
+  config_updates['WEB_CONCURRENCY'] = options[:web_concurrency]
+end
+
+# Are we changing dyno sizes?
+if options[:size] && options[:size] != current_size
+  formation_updates["size"] = options[:size]
+  # setting values to nil will unset the config value
+  config_updates['RUBY_GC_HEAP_GROWTH_MAX_SLOTS'] = default_config[options[:size]]['RUBY_GC_HEAP_GROWTH_MAX_SLOTS']
+  # only set WEB_CONCURRENCY to the default if we did not set it as an option
+  config_updates['WEB_CONCURRENCY'] ||= default_config[options[:size]]['WEB_CONCURRENCY']
+  # The direction determines the order of the commands
+  if options[:size] == 'PX'
+    # We are scaling UP. Switch to PX before we update WEB_CONCURRENCY
+    commands = [:formation, :config]
   end
-  if !options[:web_concurrency]
-    config_updates['WEB_CONCURRENCY'] = default_config[options[:size]]['WEB_CONCURRENCY']
+end
+
+# if preboot is enabled and we are scaling down to 1 dyno,
+# we must sleep 4 minutes between updating the config and scaling down
+if preboot_enabled && (options[:quantity] == '1' && current_formation["size"] != '1')
+  sleep_duration = 4
+else
+  sleep_duration = 0
+end
+
+commands.each_with_index do |command, index|
+  # sleep before second command, if necessary
+  if index == 1
+    sleep(sleep_duration)
```

Comment:
I think we only want to sleep if we actually executed the previous command (so, if we changed formations or if we committed a config change, which we don't always do)